runtime/src/iree/hal/device.h

```diff
@@ -248,6 +248,30 @@ iree_hal_device_host_allocator(iree_hal_device_t* device);
 IREE_API_EXPORT iree_hal_allocator_t* iree_hal_device_allocator(
     iree_hal_device_t* device);
 
+// Replaces the default device memory allocator.
+// The |new_allocator| will be retained for the lifetime of the device or the
+// allocator is replaced again. The common usage pattern is to shim the default
+// allocator with a wrapper:
+//   // Retain the existing allocator in the new wrapper.
+//   wrap_allocator(iree_hal_device_allocator(device), &new_allocator);
+//   // Update the device to use the wrapper for allocations.
+//   iree_hal_device_replace_allocator(device, new_allocator);
+//
+// WARNING: this is not thread-safe and must only be performed when the device
+// is idle and all buffers that may have been allocated from the existing
+// allocator have been released. In general the only safe time to call this is
+// immediately after device creation upon startup and before any buffers have
+// been allocated. Beware: there are no internal checks for this condition!
+//
+// TODO(benvanik): remove this method and instead allow allocators to be
+// composed without the safety caveats. This may take the form of unbound
+// allocators that the device can inject the base allocator into. Another
+// approach would be to replace the singular allocator with queue-specific pools
+// and make the user register those pools explicitly with the implementation
+// they desire.
+IREE_API_EXPORT void iree_hal_device_replace_allocator(
+    iree_hal_device_t* device, iree_hal_allocator_t* new_allocator);
```

Comment:
Do any of the HAL drivers allocate buffers during device creation?

I think the WebGPU driver does, at least: https://github.com/iree-org/iree/blob/05878ed58327765e0661c056557e0e60f65eb376/runtime/src/iree/hal/drivers/webgpu/webgpu_device.c#L123-L132

It would be safer to pass a callback that can be used to shim/wrap as an argument during driver/device creation. As long as some drivers don't also allocate buffers during creation, maybe the sharp edge for those that do is acceptable.

Reply:
Yeah.... you hit the exact thing that's had me stuck - this is (at least) the 5th version of this PR I've written before calling it :)
![image](https://user-images.githubusercontent.com/75337/212131611-b2881c38-b9f3-48c4-b5cc-a37a920beacd.png)

The callback approach gets practically impossible due to all the ways devices can be created (via flags, via registry, via direct calls, via wrapping, etc). There was no way I could find to avoid pulling in all possible allocators into all possible backends in every possible use of them. I think maybe in the future we could redesign allocators (the TODO in there) by adding an additional level of indirection such that new pools could be registered at any time to avoid this but have been stuck on this for a few months and just need to make progress :)

The tricky part of the problem is that we can't have buffers retain their allocators as that'd create a cycle - allocators may retain buffers but not the other way around. I'm ok with the rough edges here for now because this should not be used by anyone but binding layers and the primary pattern that makes sense is to only wrap allocators such that lifetimes are safe. The danger is someone discarding the original allocator we create with the device while buffers are still live but if the allocator the user creates is retaining that base allocator things will still be fine.

Another thing I considered for safety was always retaining the base allocator created with the device and having the allocator set by the user be separate but that has other caveats depending on what the base allocators are. For example today with the VMA allocator in Vulkan we may want to swap that out and not have VMA always initialized and retaining blocks of memory even if it's unused.

I suspect NOD will be the first folks besides me using this API and once we have some examples we can iterate some more.

Reply:
Okay, SGTM. For WebGPU maybe we could defer the staging buffer allocation to after device creation (implicit on first use, or explicit after creation).